server/src/cm_job_task.cpp

```diff
@@ -1538,19 +1542,97 @@ ts2_start_broker (nvplist *in, nvplist *out, char *_dbmt_error)
 {
   char *bname;
   T_CM_ERROR error;
+  struct {
+    long mtype;
+    char msg[IPC_MSG_SIZE];
+  } queue_msg;
+  int qid, nbytes, pid, gpid = getpid ();
+  key_t key = 33000000 + gpid;
+  int retry_count = 5;
+  enum
+  {
+    CMS_NO_ERROR = ERR_NO_ERROR,
+    CMS_ER_FORK_FAIL,
+    CMS_ER_CMS
+  } ret = CMS_NO_ERROR;
 
   if ((bname = nv_get_val (in, "bname")) == NULL)
     {
       strcpy (_dbmt_error, "broker name");
       return ERR_PARAM_MISSING;
     }
 
-  if (cm_broker_on (bname, &error) < 0)
+  memset (&queue_msg, 0, sizeof (queue_msg));
+  queue_msg.mtype = gpid;
+  qid = msgget(key, IPC_CREAT | 0600);
+
+  if ((pid = fork ()) < 0)
     {
-      strcpy (_dbmt_error, error.err_msg);
-      return ERR_WITH_MSG;
+      ret = CMS_ER_FORK_FAIL;
+      goto fin;
     }
-  return ERR_NO_ERROR;
+
+  if (pid == 0)
+    {
+      int child_pid, cm_ret;
+
+      if ((child_pid = fork ()) < 0)
+        {
+          if (qid > 0)
+            {
+	            snprintf (queue_msg.msg, IPC_MSG_SIZE - 1, "fork failed");
+              msgsnd(qid, (void *) &queue_msg, IPC_MSG_SIZE, 0);
+            }
+          exit (255);
+        }
+
+      if (child_pid)
+        {
+          exit (0);
+        }
+
+      setsid();
+      for (int i = 3; i < 1024; i++)
+        {
+          close (i);
+        }
+
+      cm_ret = cm_broker_on (bname, &error);
+      snprintf (queue_msg.msg, IPC_MSG_SIZE - 1, "%s", cm_ret < 0 ? error.err_msg : "OK");
+      msgsnd(qid, (void *) &queue_msg, IPC_MSG_SIZE, 0);
+      exit (0);
+    }
+
+  for (int i = 0; i < retry_count; i++)
+    {
+      SLEEP_MILISEC (0, 200);
+      waitpid(-1, NULL, WNOHANG);
+      if (msgrcv(qid, &queue_msg, IPC_MSG_SIZE, gpid, IPC_NOWAIT) > 0)
+        {
+          LOG_ERROR ("broker start: %s", queue_msg.msg);
+          ret = strcmp (queue_msg.msg, "OK") == 0 ? CMS_NO_ERROR : CMS_ER_CMS;
+        }
+    }
+
+fin:
+  if (qid > 0)
+    {
+      msgctl (qid, IPC_RMID, NULL);
+    }
+
+  switch (ret)
```

Comment:
If msgrcv() fails until retry_count is exceeded, it is processed as CMS_NO_ERROR, which is the default value of ret. 
Is this correct?

Reply:
It is unclear situation when cub_manager fails to receive IPC message from the grandchild process.
Probably, cub_manager fails both creating broker process and sending IPC message or
even it success to create broker process but fails to send IPC message.
Should I handle it as ERROR?

Reply:
I will handle it with ERROR message, "unknown OS error"

Reply:
how about ERROR message as "timeout error of message reception" ?

Reply:
I changed error message as:
`failed to receive IPC message from cub_manager (timeout)`